plutus-contract-exe/src/Language/Plutus/Contract/Contract.hs

```diff
@@ -0,0 +1,86 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TypeFamilies      #-}
+module Language.Plutus.Contract.Contract where
+
+import           Control.Applicative (liftA2)
+import           Control.Monad       ((>=>))
+import           Data.Bifunctor      (first)
+import           Data.List           (foldl')
+
+data Contract t i a =
+    Waiting (i -> Contract t i a)
+    | Done
+    | Emit t (Contract t i a) -- produce a 't' value
+    | Pure a
+    deriving (Functor)
+
+-- The applicative instance parallelises the 'Waiting' operations
+instance Applicative (Contract t i) where
+    pure = Pure
+    cf <*> ca = case cf of
+        Done -> Done
+        Emit t c -> Emit t (c <*> ca)
+        Pure f -> case ca of
+            Done       -> Done
+            Emit t c   -> Emit t (Pure f <*> c)
+            Pure a     -> Pure (f a)
+            Waiting f' -> Waiting $ \i' -> fmap f (f' i')
+        Waiting f ->
```

Comment:
I feel like this case could be simpler too but I'm not clever enough to see how atm.

Reply:
Hm. So having

```haskell
f :: i -> Contract t i (a -> b)
c :: Contract t i a
```

we need to produce

```haskell
res :: Contract t i b
```

We can transform `f` into this form:

```haskell
f :: i -> Contract t i a -> Contract t i b
```

and the overall thing looks like mapping a function [iover](http://hackage.haskell.org/package/lens-4.17.1/docs/Control-Lens-Setter.html#v:iover) a contract.

Alternatively, `f` is an instance (and reordering) of

```haskell
f :: (i -> Contract t i a) -> i -> Contract t i b
```

So the general signature is

```haskell
((i -> Contract t i a) -> i -> Contract t i b) -> Contract t i a -> Contract t i b
```

and morally this functions transforms a continuation of type `i -> Contract t i a` into some other one in a `Contract t i a`.

I'm not able to spot a nicer pattern.

Reply:
Here's my attempt at a rewrite, which I think is nicer.
```
    Emit t c  <*> ca       = Emit t (c <*> ca)
    cf        <*> Emit t c = Emit t (cf <*> c)
    Pure f    <*> ca       = f <$> ca
    Waiting f <*> ca       = Waiting $ \i -> f i <*> offer i ca
        where
            offer i (Waiting f) = f i
            offer _ c           = c
```

Reply:
That's good, it makes sense to look at the `Emit` on the right as well.